tests/unit/test_gce_clients.py

```diff
@@ -0,0 +1,344 @@
+
+
+class TestPagingGoogleClientBase:
+    @pytest.fixture
+    def simple_paging_client(self, mocker, get_gce_client):
+        class TestClient(gce_clients.PagingGoogleClientBase):
+            def parse_response_into_items(self, response, items):
+                items.append(response['data'])
+
+        client = get_gce_client(TestClient)
+        return client
+
+    def test_parse_response_into_items_raises_when_not_implemented(
+            self, fake_auth_client):
+        paging_base = gce_clients.PagingGoogleClientBase(fake_auth_client)
+        with pytest.raises(NotImplementedError):
+            paging_base.parse_response_into_items(None, [])
+            paging_base._session.close()
+
+    @pytest.mark.asyncio
+    async def test_list_all_items_returns_empty_list_on_failed_requests(
+            self, mocker, mock_coro, simple_paging_client, caplog):
+        caplog.set_level(logging.ERROR)
+        sleep_mock, sleep_coro = mock_coro
+        mocker.patch('gordon_janitor_gcp.gce_clients.asyncio.sleep', sleep_coro)
+        retries = 3
+        retry_wait = 5
+        with aioresponses() as m:
+            m.get(
+                'http://example.com',
+                exception=exceptions.GCPHTTPError('504 Gateway Timeout'))
+
+            results = await simple_paging_client.list_all_items(
+                'http://example.com', {},
+                retries=retries,
+                retry_wait=retry_wait)
+        assert results == []
+        expected_sleeps = [mocker.call(retry_wait) for _ in range(retries)]
+        sleep_mock.assert_has_calls(expected_sleeps)
+        assert len(caplog.records) == 4
+
+    @pytest.mark.asyncio
+    async def test_list_all_items_returns_some_items_on_some_failed_requests(
+            self, mocker, mock_coro, simple_paging_client, caplog):
+        caplog.set_level(logging.ERROR)
+        sleep_mock, sleep_coro = mock_coro
+        mocker.patch('gordon_janitor_gcp.gce_clients.asyncio.sleep', sleep_coro)
+        retries = 3
+        retry_wait = 5
+        with aioresponses() as m:
+            # One successful call
+            m.get(
+                'http://example.com',
+                payload={'data': 'data', 'nextPageToken': 'token'})
+            # second call fails despite retries
+            m.get(
+                'http://example.com/?pageToken=token',
+                exception=exceptions.GCPHTTPError('504 Gateway Timeout'))
+
+            results = await simple_paging_client.list_all_items(
+                'http://example.com', {},
+                retries=retries,
+                retry_wait=retry_wait)
+        assert results == ['data']
+        expected_sleeps = [mocker.call(retry_wait) for _ in range(retries)]
+        sleep_mock.assert_has_calls(expected_sleeps)
+        assert len(caplog.records) == 4
+
+
+class TestCRMClient:
+    @pytest.fixture
+    def crm_base_url(self):
+        return 'https://cloudresourcemanager.googleapis.com/v1/projects?'
```

Comment:
I would not use the real URL for fear of accidentally improperly mocking it out and making a real call.

Reply:
Great catch! I've changed it so it's a fake url and the `CRMClient.BASE_URL` attribute gets patched with it.